cvmfs/catalog_diff_tool_impl.h

```diff
@@ -76,8 +77,68 @@ bool CatalogDiffTool<RoCatalogMgr>::Init() {
 }
 
 template <typename RoCatalogMgr>
-bool CatalogDiffTool<RoCatalogMgr>::Run(const PathString& path) {
-  DiffRec(path);
+bool CatalogDiffTool<RoCatalogMgr>::FastPathDiff(const PathString& lease_path_r) {
+  bool success = false;
+
+  catalog::DirectoryEntry old_dirent_parent;
+  catalog::DirectoryEntry old_dirent;
+  catalog::DirectoryEntry new_dirent;
+
+  bool has_old_dirent_parent;
+  bool has_old_dirent;
+  bool has_new_dirent;
+
+  PathString lease_path= PathString("/" + lease_path_r.ToString());
+  PathString lease_path_parent = GetParentPath(lease_path);
+
+  has_old_dirent_parent = old_catalog_mgr_->LookupPath( lease_path_parent, catalog::kLookupDefault, &old_dirent_parent);
+  has_old_dirent        = old_catalog_mgr_->LookupPath( lease_path, catalog::kLookupDefault, &old_dirent);
+  has_new_dirent        = new_catalog_mgr_->LookupPath( lease_path, catalog::kLookupDefault, &new_dirent);
+
+  if( !has_old_dirent_parent || !has_new_dirent) { return false; }
+
+  if(   has_old_dirent && old_dirent.IsDirectory() && old_dirent.IsNestedCatalogMountpoint()
+      && new_dirent.IsDirectory() && new_dirent.IsNestedCatalogMountpoint() ) {
+     // switching from nested catalog to nested catalog
+      FileChunkList chunks;
+      XattrList xattrs;
+      if (new_dirent.HasXattrs()) {
+        new_catalog_mgr_->LookupXattrs(lease_path, &xattrs);
+      }
+      LogCvmfs(kLogCvmfs, kLogSyslog, "Replacing existing nested catalog for directory %s", lease_path.c_str());
+
+      ReportModification(lease_path, old_dirent, new_dirent, xattrs, chunks);
```

Comment:
the original `DiffRec` implementation has a quick escape when old and new dirent are identical.
not sure how much it safes time-wise?

https://github.com/cvmfs/cvmfs/blob/19ebfba81b765e9be0deff532fd0234b02382f21/cvmfs/catalog_diff_tool_impl.h#L189-L199